velox/functions/sparksql/DecimalUtil.h

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <boost/multiprecision/cpp_int.hpp>
+
+#include "velox/type/DecimalUtil.h"
+#include "velox/type/Type.h"
+
+namespace facebook::velox::functions::sparksql {
+using uint128_t = __uint128_t;
+using int256_t = boost::multiprecision::int256_t;
+
+class DecimalUtil {
+ public:
+  template <class T, typename = std::enable_if_t<std::is_same_v<T, int64_t>>>
+  inline static int64_t convert(int256_t in, bool& overflow) {
+    int64_t result = 0;
+    constexpr int256_t uint64Mask =
+        static_cast<int256_t>(std::numeric_limits<uint64_t>::max());
+
+    int256_t inAbs = abs(in);
+    bool isNegative = in < 0;
+
+    uint64_t unsignResult = (inAbs & uint64Mask).convert_to<uint64_t>();
+    inAbs >>= 64;
+
+    if (inAbs > 0) {
+      // We've shifted in by 128-bit, so nothing should be left.
+      overflow = true;
+    } else if (unsignResult > INT64_MAX) {
+      // The high-bit must not be set (signed 128-bit).
+      overflow = true;
+    } else {
+      result = static_cast<int64_t>(unsignResult);
+    }
+    if (result > velox::DecimalUtil::kShortDecimalMax) {
+      overflow = true;
+    }
+    return isNegative ? -result : result;
+  }
+
+  template <class T, typename = std::enable_if_t<std::is_same_v<T, int128_t>>>
+  inline static int128_t convert(int256_t in, bool& overflow) {
+    int128_t result = 0;
+    constexpr int128_t int128Max = int128_t(-1L) >> 1;
+    constexpr int256_t uint128Mask = std::numeric_limits<uint128_t>::max();
+
+    int256_t inAbs = abs(in);
+    bool isNegative = in < 0;
+
+    uint128_t unsignResult = (inAbs & uint128Mask).convert_to<uint128_t>();
```

Comment:
If possible these 2 can be merged.

Reply:
No, T is int64_t while the variable type uses uint64_t, we cannot get template UT as uint64_t

Reply:
Does it require templating? You are using specific types in the functions. You could simply overload the functions.
or
May be you can still merge convert<int64_t> and convert<int128_t>.

Do 
```
typedef typename std::conditional<std::is_same_v<TInput, int64_t>,uint64_t,__uint128_t>::type UT;

```
You can switch the unsigned type depending on it.

Reply:
Thanks for your guide, fixed.